### ISSUE TITLE
Static nullroutes does not show up in routing table in 23.7

### ISSUE THREAD

**Issue by joachimtingvold** [opnsense/plugins#3520]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Describe the bug**

 * Upgraded from 23.1.11_1 to 23.7. Using BGP via FRR. 
 * Redistributing static nullroutes (System -> Routes -> Configuration) worked in 23.1, but not in 23.7.
 * Static nullroutes does not show at all in FRR (Routing -> Diagnostics -> General -> IPv4/IPv6 routes).
 * Static nullroutes does not show at all in OPNsense (System -> Routes -> Status)
 * Static nullroutes *do* show up via `netstat -rn`
 * Happens to both IPv4 and IPv6 static nullroutes.
 * The new redistribution option "Statically configured routes" has no effect (i.e. selecting that, the routes still don't show, even though that technically only should affect what routes BGP learns, and not what FRR has in its RIB).
 * Disabling the static nullroutes, and re-enabling them does not work.
 * Deleting the static nullroutes, and re-adding them does not work.

**To Reproduce**

1. Add static nullroutes, both IPv4 and IPv6 (System -> Routes -> Configuration)
2. Static nullroutes should show in routing table, but does not (Routing -> Diagnostics -> General)

**Expected behavior**

Static nullroutes should show up in routing table.

**Describe alternatives you considered**

Unchecking "Network Import-Check" and adding the nullroutes as Networks in the BGP-config is our current workaround to get this working.

**Screenshots**

N/A

**Relevant log files**

N/A

**Additional context**

N/A

**Environment**

OPNsense 23.7-amd64
Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (20 cores, 40 threads)
Network Intel X520-DA2

**Comment by mimugmail** [comment]
@AdSchellevis I guess this is due to the API diagnostics rewrite as the routes pop up via netstat?

**Comment by AdSchellevis** [comment]
@mimugmail I wouldn't expect so, but it should be easy to inspect the data as frr reports it. @joachimtingvold can you check the vtysh output of the following commands:

```
/usr/local/bin/vtysh -c 'show bgp ipv4'
/usr/local/bin/vtysh -c 'show ip route'

/usr/local/bin/vtysh -c 'show ip route json'
/usr/local/bin/vtysh -c 'show bgp ipv4 json'
```

(we use the last two commands)

**Comment by joachimtingvold** [comment]
@AdSchellevis;

```foo@bar-fw1:~ # netstat -rn | grep 10.1.1
10.1.1.0/24        127.0.0.1          USB         lo0
foo@bar-fw1:~ # /usr/local/bin/vtysh -c 'show bgp ipv4 10.1.1.0/24'
% Network not in table
foo@bar-fw1:~ # /usr/local/bin/vtysh -c 'show ip route 10.1.1.0/24'
% Network not in table
foo@bar-fw1:~ # /usr/local/bin/vtysh -c 'show bgp ipv4'|grep 10.1.1
foo@bar-fw1:~ # /usr/local/bin/vtysh -c 'show ip route' | grep 10.1.1```

**Comment by mimugmail** [comment]
Can you post the output of "service frr restart" via CLI?

**Comment by joachimtingvold** [comment]
```
foo@bar-fw1:~ # service frr restart
Stopping bgpd.
Waiting for PIDS: 17465.
Stopping bfdd.
Waiting for PIDS: 14152.
Stopping zebra.
Waiting for PIDS: 8165.
Checking zebra.conf
2023/08/04 07:22:48 ZEBRA: [NNACN-54BDA][EC 4043309110] Disabling MPLS support (no kernel support)
OK
Starting zebra.
2023/08/04 07:22:48 ZEBRA: [NNACN-54BDA][EC 4043309110] Disabling MPLS support (no kernel support)
Checking bfdd.conf
OK
Starting bfdd.
Checking bgpd.conf
Graceful restart configuration changed, reset all peers to take effect
OK
Starting bgpd.
foo@bar-fw1:~ #
```

**Comment by mimugmail** [comment]
Hm, so it's something inside frr8 which might changed some defaults (again). They did when going from 7.3 to 7.4 which broke stuff already.

**Comment by tatsushid** [comment]
Hi,

I met this issue too and tried to find out what happened. I'd like to share the result for further discussion and finding a solution while this might include a mistake as I'm an OPNsense newbie and not much familiar with FreeBSD world so appreciate correcting a mistake if you find it

In summary, I believe this issue comes from an internal change between FreeBSD 13.0 and 13.2 and FRRouting internal behavior mismatch. Here is the detail. All these are tested on OPNsense 24.1 but the basics should be same on 23.7 I think

First, at adding a null route from OPNsense "System -> Routes -> Configuration", it does same as running a command like

```
route add -net <prefix> -gateway 127.0.0.1 -blackhole
```

It adds a route with some flags. For example, when I run the command with a prefix 10.11.12.0/30 and `-v` option, it shows

```
root@OPNsense:~ # route -v add -net 10.11.12.0/30 -gateway 127.0.0.1 -blackhole
RTA_DST: inet 10.11.12.0; RTA_GATEWAY: inet 127.0.0.1; RTA_NETMASK: inet 255.255.255.252; RTM_ADD: Add Route: len 200, pid: 0, seq 1, errno 0, flags:<UP,GATEWAY,STATIC,BLACKHOLE>
locks:  inits:
sockaddrs: <DST,GATEWAY,NETMASK>
 10.11.12.0 localhost 255.255.255.252
add net 10.11.12.0: gateway 127.0.0.1
```

As you can see, the command tries to add the route with UP, GATEWAY, STATIC, BLACKHOLE flags and at the same time, I run `route -v -n monitor` to watch what is added, it shows

```
root@OPNsense:~ # route -v -n monitor

got message of size 200 on Thu Jul  4 23:47:08 2024
RTM_ADD: Add Route: len 200, pid: 31929, seq 1, errno 0, flags:<UP,DONE,STATIC,BLACKHOLE>
locks:  inits:
sockaddrs: <DST,GATEWAY,NETMASK>
 10.11.12.0 127.0.0.1 255.255.255.252
```

It looks that the route is added with UP, STATIC, BLACKHOLE (DONE should be just telling the operation has finished, not a permanent flag I think). These flags can also be seen in `netstat -nr` result like

```
root@OPNsense:~ # netstat -nr | grep -F 10.11.12.0/30
10.11.12.0/30      127.0.0.1          USB         lo0
```

So it also shows the route has flags UP, STATIC, BLACKHOLE

However, as it is commented at [the forum topic](https://forum.opnsense.org/index.php?topic=35763.0), this kind of route had flags like UGSB (UP, GATEWAY, STATIC, BLACKHOLE) on OPNsense 23.1 system. It was based on FreeBSD 13.0 while OPNsense 23.7 and 24.1 are based on FreeBSD 13.2. Between these versions, something was changed

After digging into FreeBSD codes, I found this commit
https://github.com/freebsd/freebsd-src/commit/d2deebe21b591336fbd8915b37d409b25da54d4d

This drops GATEWAY flag when BLACKHOLE flag is set. This change seems to have been introduced from 13.2 and didn't exist until 13.1. It matches with the route flag result above

On the other hand, how does FRRouting import FreeBSD system route into its routing table? A system route comes through Zebra component of FRRouting and is added as a kernel route in most cases (the other case is direct type). It is read through [`rtm_read`](https://github.com/FRRouting/frr/blob/d4758b3ccc76c77d0b8ba255089f79e67379bb30/zebra/kernel_socket.c#L991) function and you can easily find [these lines](https://github.com/FRRouting/frr/blob/d4758b3ccc76c77d0b8ba255089f79e67379bb30/zebra/kernel_socket.c#L1032-L1034) in the function. Yes, FRRouting simply ignores a route without GATEWAY flag. This is a really long live code as commented in https://github.com/FRRouting/frr/issues/9185

In conclusion, I think the issue should come from FreeBSD internal change and it hits FRRouting really old route rejection behavior

But from the result, I couldn't find a good solution for it. Fixing FRRouting looks being stuck from the issue above and similar cases there. Changing FreeBSD itself again is not much realistic only for one software from my view, neither. Removing `-blackhole` flag at adding a null route keeps GATEWAY flag and the route is imported into FRRouting but it may not be "null" route

Hope someone will give a nice suggestion on this
